### PR TITLE
Plot improvements

### PR DESCRIPTION
**Pull request by AlexanderFengler**
PR for plot improvements.

- Main feature addition will be `model_cartoon` plots.
- Small improvements to the pre-existing plots.

### CODE REVIEW

**Comment by AlexanderFengler** [comment]
This is still work in progress, roughly 80% done.

**Comment by cpaniaguam** [comment]
@AlexanderFengler What do you think about creating a PR (ideally a few given its size) for `model_cartoon.py`? I think this will facilitate reviewing.

**Comment by AlexanderFengler** [comment]
> @AlexanderFengler What do you think about creating a PR (ideally a few given its size) for `model_cartoon.py`? I think this will facilitate reviewing.

Sorry I know this is a big PR, but I really can't go back an disentangle this easily now.
Will avoid (and generally do) big PRs like this.
Changes are all interrelated I can't make a PR out of only the changes to `model_cartoon.py` here, it would take a lot of effort.

`model_cartoon.py` is new and the major piece, the rest is just a bunch of refactor, some little utilities, and minor subtle bug fixes in the other plotting functions (me looking at those closely was overdue).

The ++ line count is also inflated by having added two data files, so it's not as bad as it seems on first glance :).

File: tests/test_plotting_cartoon.py [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+        (2, None, True, False, "participant_id", "stim"),
+        (2, None, True, False, None, None),
+    ],
+)
+def test_plot_model_cartoon_2_choice(
+    cav_model_cartoon, n_trajectories, groups, plot_pp_mean, plot_pp_samples, row, col
+):
+    """Test plot_model_cartoon for 2-choice data."""
+    if (not plot_pp_mean) and (not plot_pp_samples):
+        with pytest.raises(AssertionError):
```
Comment (digicosmos86) [inline tests/test_plotting_cartoon.py:54]: Again, I think `ValueError` conveys what the type of the error is and might be a better fit here

File: tests/test_plotting_cartoon.py [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+            n_samples=10,
+            bins=30,
+            col=col,
+            row=row,
+            groups=groups,
+            plot_pp_mean=plot_pp_mean,
+            plot_pp_samples=plot_pp_samples,
+            alpha=0.025,
+            **{"n_trajectories": n_trajectories},
+        )
```
Comment (digicosmos86) [inline tests/test_plotting_cartoon.py:79]: Should we actually test something here (like, dimensions of graphs, etc), so it's not just making sure the code runs?

File: src/hssm/plotting/model_cartoon.py [hunk 3]
```diff
@@ -0,0 +1,1985 @@
...
+    # ADD MODEL:
+
+    # RUN SIMULATIONS
+    # -------------------------------
+
+    # Simulator Data
+    if random_state is not None:
+        np.random.seed(random_state)
+
+    rand_int = np.random.choice(400000000)
```
Comment (digicosmos86) [inline src/hssm/plotting/model_cartoon.py:1462]: Separate this constant too?

File: src/hssm/plotting/model_cartoon.py [hunk 4]
```diff
@@ -0,0 +1,1985 @@
...
+                draw_tmp = np.random.choice(
+                    theta_posterior.index.get_level_values("draw")
+                )
+                obs_tmp = np.random.choice(
+                    theta_posterior.index.get_level_values("obs_n")
+                )
+
+                tmp_theta = theta_posterior.loc[
+                    (chain_tmp, draw_tmp, obs_tmp), :
+                ].values
```
Comment (digicosmos86) [inline src/hssm/plotting/model_cartoon.py:1539]: Maybe use a loop or a tuple comprehension to avoid repetition?

**Review by digicosmos86 — APPROVED** [review]
Wasn't able to go through everything but looks good in principle. Minor comments in the review

File: src/hssm/plotting/posterior_predictive.py [hunk 5]
```diff
@@ -189,21 +191,98 @@ def _plot_posterior_predictive_2D(
...
+        g.figure.subplots_adjust(top=0.9)
+        g.figure.suptitle(title)
 
     g.set_xlabels(xlabel)
     g.set_ylabels(ylabel)
 
     return g
 
 
+def _process_linestyles_pp(linestyles):
```
Comment (cpaniaguam) [inline src/hssm/plotting/posterior_predictive.py:209]:
```suggestion
def _process_linestyles_pp(linestyles: str | Iterable):
```

File: src/hssm/plotting/posterior_predictive.py [hunk 6]
```diff
@@ -189,21 +191,98 @@ def _plot_posterior_predictive_2D(
...
+    if isinstance(linestyles, str):
+        return [linestyles, linestyles]
```
Comment (cpaniaguam) [inline src/hssm/plotting/posterior_predictive.py:211]: Little less code.
```suggestion
        return [linestyles] * 2
```

File: src/hssm/plotting/posterior_predictive.py [hunk 7]
```diff
@@ -189,21 +191,98 @@ def _plot_posterior_predictive_2D(
...
+                return linestyles
+            else:
+                raise ValueError(
+                    "The `linestyles` argument must be a string or a list of length 1 "
+                    "or 2."
+                )
+        else:
+            raise ValueError(
+                "The `linestyles` argument must be a string or a list of strings."
+            )
```
Comment (cpaniaguam) [inline src/hssm/plotting/posterior_predictive.py:227]: Where do you stand on guard clauses?
```suggestion
    if isinstance(linestyles, (list, tuple)):
        linestyles = list(linestyles)
        if not all(isinstance(ls, str) for ls in linestyles):
            raise ValueError("The `linestyles` list must contain only strings.")
        if len(linestyles) in {1, 2}:
            return linestyles * 2 if len(linestyles) == 1 else linestyles
        raise ValueError("The `linestyles` list must have length 1 or 2.")
```

File: src/hssm/plotting/posterior_predictive.py [hunk 8]
```diff
@@ -189,21 +191,98 @@ def _plot_posterior_predictive_2D(
...
+            raise ValueError(
+                "The keys of the `linestyles` dictionary must be 'predicted' and/or "
+                "'observed'."
+            )
+        linestyles = [linestyles.get("predicted", "-"), linestyles.get("observed", "-")]
+    else:
+        raise ValueError(
+            "The `linestyles` argument must be a string, a list of strings,"
+            " or a dictionary."
+        )
```
Comment (cpaniaguam) [inline src/hssm/plotting/posterior_predictive.py:240]:
```suggestion
    if not isinstance(linestyles, dict):
        raise ValueError(
            "The `linestyles` argument must be a dictionary."
        )

    allowed_keys = {"predicted", "observed"}
    if not set(linestyles) <= allowed_keys:
        raise ValueError("Keys must be 'predicted' and/or 'observed'.")
        
    linestyles = [linestyles.get(key, "-") for key in allowed_keys]
```

File: src/hssm/plotting/posterior_predictive.py [hunk 9]
```diff
@@ -189,21 +191,98 @@ def _plot_posterior_predictive_2D(
...
+        linewidths = [
+            linewidths.get("predicted", 1.25),
+            linewidths.get("observed", 1.25),
+        ]
+    else:
+        raise ValueError(
+            "The `linewidths` argument must be a float, a list of floats,"
+            "or a dictionary."
+        )
+    return linewidths
```
Comment (cpaniaguam) [inline src/hssm/plotting/posterior_predictive.py:279]: This function's logic is almost identical as that of `_process_linestyles_pp`'s. Could a more abstract one be defined and called in either of them?

File: src/hssm/plotting/posterior_predictive.py [hunk 10]
```diff
@@ -343,6 +427,17 @@ def plot_posterior_predictive(
     mpl.axes.Axes | sns.FacetGrid
         The matplotlib `axis` or seaborn `FacetGrid` object containing the plot.
     """
+    # Process hdi
+    if hdi is not None:
+        interval = _hdi_to_interval(hdi=hdi)
+    else:
+        interval = None
```
Comment (cpaniaguam) [inline src/hssm/plotting/posterior_predictive.py:434]:
```suggestion
    interval = _hdi_to_interval(hdi=hdi) if hdi is not None else None
```

**Review by cpaniaguam — CHANGES_REQUESTED** [review]
Only looked at `posterior_predictive.py`. I'll probably have more to say about the others later.

File: src/hssm/plotting/model_cartoon.py [hunk 11]
```diff
@@ -0,0 +1,1985 @@
...
+            reps=sim_out["rts"][
+                (sim_out["rts"] != -999) & (sim_out["choices"] == 1)
+            ].shape[0],
+        )
+        weights_down = np.tile(
+            (1 / bin_size) / sim_out["rts"][(sim_out["rts"] != -999)].shape[0],
+            reps=sim_out["rts"][
+                (sim_out["rts"] != -999) & (sim_out["choices"] != 1)
+            ].shape[0],
+        )
```
Comment (cpaniaguam) [inline src/hssm/plotting/model_cartoon.py:893]:
```suggestion
    if theta_mean is not None:
        valid_mask = sim_out["rts"] != -999
        valid_rts, valid_choices = (sim_out[key][valid_mask] for key in ["rts", "choices"])
        weight_base = (1 / bin_size) / valid_rts.shape[0]
        valid_choices = valid_choices == 1
        weights_up, weights_down = [
            np.tile(weight_base, reps=valid_rts[vc].shape[0]) for vc in [valid_choices, ~valid_choices]
        ]
```

File: src/hssm/plotting/model_cartoon.py [hunk 12]
```diff
@@ -0,0 +1,1985 @@
...
+            bins=bins,
+            weights=weights_down,
+            histtype=hist_histtype,
+            bottom=hist_bottom_low,
+            alpha=1,
+            color=color_pp_mean,
+            edgecolor=color_pp_mean,
+            linewidth=linewidth_histogram,
+            zorder=-1,
+        )
```
Comment (cpaniaguam) [inline src/hssm/plotting/model_cartoon.py:924]: Here the common settings could be extracted or reuse the values suggested earlier to make code a bit more DRY.

File: src/hssm/plotting/model_cartoon.py [hunk 13]
```diff
@@ -0,0 +1,1985 @@
...
+                bins=bins,
+                weights=weights_down,
+                histtype=hist_histtype,
+                bottom=hist_bottom_low,
+                alpha=alpha_pp,
+                color=color_pp,
+                edgecolor=color_pp,
+                linewidth=linewidth_histogram,
+                zorder=(-k - 1),
+            )
```
Comment (cpaniaguam) [inline src/hssm/plotting/model_cartoon.py:977]: This is almost the same logic as above. Consider moving this to a helper function and reuse it.

File: src/hssm/plotting/model_cartoon.py [hunk 14]
```diff
@@ -0,0 +1,1985 @@
...
+            bins=bins,
+            weights=weights_down_data,
+            histtype=hist_histtype,
+            bottom=hist_bottom_low,
+            alpha=1,
+            color=color_data,
+            edgecolor=color_data,
+            linewidth=linewidth_histogram,
+            zorder=-1,
+        )
```
Comment (cpaniaguam) [inline src/hssm/plotting/model_cartoon.py:1018]: Reuse here also.

File: src/hssm/plotting/model_cartoon.py [hunk 15]
```diff
@@ -0,0 +1,1985 @@
...
+    **kwargs
+        Additional keyword arguments passed to plotting functions.
+    """
+    # Check markercolor type
+    if isinstance(markercolor_trajectory_rt_choice, str):
+        markercolor_trajectory_rt_choice_dict = {}
+        for value_ in sample[0]["metadata"]["possible_choices"]:
+            markercolor_trajectory_rt_choice_dict[value_] = (
+                markercolor_trajectory_rt_choice
+            )
```
Comment (cpaniaguam) [inline src/hssm/plotting/model_cartoon.py:1145]: Use a dict comprehension.
```suggestion
        markercolor_trajectory_rt_choice_dict = {
            value_: markercolor_trajectory_rt_choice
            for value_ in sample[0]["metadata"]["possible_choices"]
        }
```

File: src/hssm/plotting/model_cartoon.py [hunk 16]
```diff
@@ -0,0 +1,1985 @@
...
+        cnt = 0
+        for value_ in sample[0]["metadata"]["possible_choices"]:
+            markercolor_trajectory_rt_choice_dict[value_] = (
+                markercolor_trajectory_rt_choice[cnt]
+            )
+            cnt += 1
+    elif isinstance(markercolor_trajectory_rt_choice, dict):
+        markercolor_trajectory_rt_choice_dict = markercolor_trajectory_rt_choice
+    else:
+        pass
```
Comment (cpaniaguam) [inline src/hssm/plotting/model_cartoon.py:1156]: What would happen if none of the above gets evaluated?

File: src/hssm/plotting/model_cartoon.py [hunk 17]
```diff
@@ -0,0 +1,1985 @@
...
+            color_trajectories_dict[value_] = color_trajectories
+    elif isinstance(color_trajectories, list):
+        cnt = 0
+        for value_ in sample[0]["metadata"]["possible_choices"]:
+            color_trajectories_dict[value_] = color_trajectories[cnt]
+            cnt += 1
+    elif isinstance(color_trajectories, dict):
+        color_trajectories_dict = color_trajectories
+    else:
+        pass
```
Comment (cpaniaguam) [inline src/hssm/plotting/model_cartoon.py:1171]: Consider the patterns suggested above.

File: src/hssm/plotting/model_cartoon.py [hunk 18]
```diff
@@ -0,0 +1,1985 @@
...
+            if highlight_rt_choice and tmp_traj_choice == j:
+                axis.scatter(
+                    t_s[tmp_maxid] + sample[i]["metadata"]["t"][0],
+                    b_tmp,
+                    marker_size_rt_choice,
+                    color=color_dict[tmp_traj_choice],
+                    alpha=1,
+                    marker=marker_type_rt_choice,
+                    zorder=2000 + i,
+                )
```
Comment (cpaniaguam) [inline src/hssm/plotting/model_cartoon.py:1869]: Consider defining some names and reusing them. For example, `metadata = sample[i]["metadata"]`.

**Review by cpaniaguam — COMMENTED** [review]
Thanks for this! I pointed out a number of opportunities for improvement. A few functions seem to be quite similar and perhaps more general helper functions could be defined and reused.

File: src/hssm/plotting/model_cartoon.py [hunk 19]
```diff
@@ -0,0 +1,1985 @@
...
+
+    deterministics_idata = xr.merge(deterministics_list)
+    setattr(idata, "posterior", xr.merge([idata.posterior, deterministics_idata]))
+    return idata
+
+
+def attach_trialwise_params_to_df(model, df, idata):
+    """Attach the trial-wise parameters to the dataframe."""
+    necessary_params = default_model_config[model.model_name]["list_params"]
```
[the remark this answers is not in the document]
Comment (AlexanderFengler) [reply src/hssm/plotting/model_cartoon.py:251]: that's right, for now we can't support custom models, we need to define a bit more templating to make that a reality.

The cartoon plots depends on output formats coming out of `ssm-simulators` atm, which is a bit richer than basic simulator signature for random variables.

File: src/hssm/plotting/model_cartoon.py [hunk 20]
```diff
@@ -0,0 +1,1985 @@
...
+    plot_pp_samples: bool = False,
+    colors: str | list[str] | None = None,
+    linestyles: str | list[str] | tuple[str] | Dict[str, str] = "-",
+    linewidths: float | list[float] | tuple[float] | Dict[str, float] = 1.25,
+    title: str | None = "Posterior Predictive Distribution",
+    xlabel: str | None = "Response Time",
+    ylabel: str | None = "",
+    grid_kwargs: dict | None = None,
+    **kwargs,
+) -> Axes | sns.FacetGrid | list[sns.FacetGrid]:
```
[the remark this answers is not in the document]
Comment (AlexanderFengler) [reply src/hssm/plotting/model_cartoon.py:304]: don't actually want to return that dropped.

File: src/hssm/plotting/model_cartoon.py (hunk 3 above)
Comment (AlexanderFengler) [reply src/hssm/plotting/model_cartoon.py:1462]: what do you mean here?

File: tests/conftest.py [hunk 21]
```diff
@@ -99,3 +99,69 @@ def posterior():
...
+                    "Intercept": {"name": "Normal", "mu": 0.0, "sigma": 1.5},
+                },
+                "formula": "v ~ 1 + stim",
+                "link": "identity",
+            },
+            {
+                "name": "a",
+                "prior": {
+                    "Intercept": {"name": "Normal", "mu": 1.5, "sigma": 0.5},
+                },
```
[the remark this answers is not in the document]
Comment (AlexanderFengler) [reply tests/conftest.py:124]: "prior" can have multiple keys inside actually. 

E.g. could have defined the prior also for `1(|participant_id)`, but using defaults here (applied automatically upon omission).

File: tests/test_plotting_cartoon.py (hunk 2 above)
Comment (AlexanderFengler) [reply tests/test_plotting_cartoon.py:79]: added a few checks, but we might have to think up more. 
It's just an ax object, what should one generically inspect there?
Went for row_names etc. now.

File: src/hssm/plotting/model_cartoon.py (hunk 3 above)
Comment (digicosmos86) [reply src/hssm/plotting/model_cartoon.py:1521]: You can also define a constant for 400000000 here and reuse it across the file

File: src/hssm/plotting/model_cartoon.py [hunk 22]
```diff
@@ -0,0 +1,1985 @@
...
+    bins = list(np.arange(xlim_low, xlim_high, bin_size))
+
+    # RUN SIMULATIONS
+    # -------------------------------
+
+    # Simulator Data from posterior mean
+    if random_state is not None:
+        np.random.seed(random_state)
+
+    rand_int = np.random.choice(400000000)
```
[the remark this answers is not in the document]
Comment (digicosmos86) [reply src/hssm/plotting/model_cartoon.py:718]: np.random.choice is orders of magnitude slower than np.random.randint

**Review by digicosmos86 — APPROVED** [review]
Approved in principle. I see that there are some unresolved comments in model_cartoon.py. Is that because GitHub hides the file by default due to size?

**Comment by AlexanderFengler** [comment]
> Approved in principle. I see that there are some unresolved comments in model_cartoon.py. Is that because GitHub hides the file by default due to size?

hm will give it another looks, but couldn't find any more unadressed comments las time I checked.

File: src/hssm/plotting/posterior_predictive.py [hunk 23]
```diff
@@ -378,13 +485,13 @@ def plot_posterior_predictive(
...
-    if np.any(plotting_df["response"] == 0):
+    if np.any(plotting_df["response"] == 0) and model.n_choices == 2:
         plotting_df["response"] = np.where(plotting_df["response"] == 0, -1, 1)
     if model.n_choices == 2:
         plotting_df["rt"] = plotting_df["rt"] * plotting_df["response"]
 
-    # Then, plot the posterior predictive distribution against the observed data
-    # Determine whether we are producing a single plot or a grid of plots
+        # Then, plot the posterior predictive distribution against the observed data
+        # Determine whether we are producing a single plot or a grid of plots
```
[the remark this answers is not in the document]
Comment (AlexanderFengler) [reply src/hssm/plotting/posterior_predictive.py:494]: That's possible yeah. 
I actually use 4 spaces.

**Comment by AlexanderFengler** [comment]
@digicosmos86 ok to merge?

**Comment by digicosmos86** [comment]
> @digicosmos86 ok to merge?

I still see some comments in model_cartoon.py that are not resolved. Not sure if you saw them or not since large files are not rendered on Github by default. I've approved none-the-less so OK with me on my end

File: src/hssm/plotting/model_cartoon.py [hunk 24]
```diff
@@ -0,0 +1,1985 @@
...
+                else None
+            ),
+            theta_posterior=(
+                data_posterior_predictive[model_params] if plot_samples else None
+            ),
+            data=(data_observed.reset_index() if plot_data else None),
+            linewidth_histogram=kwargs.get("linewidth_histogram", 0.5),
+            linewidth_model=kwargs.get("linewidth_model", 0.5),
+            **kwargs,
+        )
```
[the remark this answers is not in the document]
Comment (AlexanderFengler) [reply src/hssm/plotting/model_cartoon.py:142]: made the change, but had to go through more mypy pain as a result :(

**Review by AlexanderFengler — COMMENTED** [review]
ok now I think everything is resolved.
Thanks for the thorough review guys.

**Pull request merged**